atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampBoundStore.java

```diff
@@ -120,67 +129,87 @@ public synchronized void storeUpperLimit(final long limit) {
         clientPool.runWithRetry(new FunctionCheckedException<Client, Void, RuntimeException>() {
             @Override
             public Void apply(Client client) {
-                cas(client, currentLimit, limit);
+                cas(client, makeNewColumn(currentLimit), currentLimit, limit);
                 return null;
             }
         });
     }
 
-    private void cas(Client client, Long oldVal, long newVal) {
-        final CASResult result;
+    private void cas(Client client, Column oldColumn, Long oldVal, long newVal) {
+        CASResult result = updateTimestampInDb(client, oldColumn, oldVal, newVal);
+        if (!result.isSuccess()) {
+            if (result.getCurrent_values().size() == 0) {
+                log.error("[CAS] Error trying to set timestamp limit to {}: there is no limit stored in DB.", newVal);
+                throw new Error("Error trying to set timestamp limit: there is no limit stored in DB.");
+            }
+            IdAndTimestamp currentIdAndTimestamp = new IdAndTimestamp(result);
+            if (currentIdAndTimestamp.hasId() && currentIdAndTimestamp.getId() == id) {
+                cas(client, makeNewColumn(currentIdAndTimestamp.getTimestamp()),
```

Comment:
I find this recursion slightly odd. It's also hard to tell which case we're handling here - it would help to extract the boolean checker to a function whose name explains what case we're in. Or, if the explanation is too long to fit in a name, add a comment.